pkg/api/graph.go

```diff
@@ -470,22 +472,65 @@ func mergeSteps(into, from CIOperatorStepWithDependencies) CIOperatorStepWithDep
 	if into.Failed == nil {
 		into.Failed = from.Failed
 	}
+	if into.Substeps == nil {
+		into.Substeps = from.Substeps
+	}
 
 	return into
 }
 
-type CIOperatorStepWithDependencies struct {
+type CIOperatorStepDetails struct {
+	CIOperatorStepDetailInfo `json:",inline"`
+	Substeps                 []CIOperatorStepDetailInfo `json:"substeps,omitempty"`
+}
+
+type CIOperatorStepDetailInfo struct {
 	StepName     string                     `json:"name"`
 	Description  string                     `json:"description"`
 	Dependencies []string                   `json:"dependencies"`
 	StartedAt    *time.Time                 `json:"started_at"`
 	FinishedAt   *time.Time                 `json:"finished_at"`
 	Duration     *time.Duration             `json:"duration,omitempty"`
-	Manifests    []ctrlruntimeclient.Object `json:"manifests"`
+	Manifests    []ctrlruntimeclient.Object `json:"manifests,omitempty"`
 	LogURL       string                     `json:"log_url,omitempty"`
 	Failed       *bool                      `json:"failed,omitempty"`
 }
 
+func (c *CIOperatorStepDetailInfo) UnmarshalJSON(data []byte) error {
```

Comment:
If someone can come up with a less inefficient way to do this, that would be great